python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py

```diff
@@ -0,0 +1,185 @@
+)
+
+DEFAULT_MATCH_DIST = 31
+DIMENSIONALITY = 256
+
+PDQIndexMatch = IndexMatchUntyped[SignalSimilarityInfoWithIntDistance, IndexT]
+
+
+class PDQIndex2(SignalTypeIndex[IndexT]):
+    """
+    Wrapper around the pdq faiss index lib using PDQMultiHashIndex
+    """
+
+    def __init__(
+        self,
+        threshold: int = DEFAULT_MATCH_DIST,
+        index: t.Optional[faiss.Index] = None,
+        entries: t.Iterable[t.Tuple[str, IndexT]] = (),
+    ) -> None:
+        super().__init__()
+        self.threshold = threshold
+
+        if index is None:
+            index = faiss.IndexFlatL2(DIMENSIONALITY)
+        self.index = _PDQHashIndex(index)
+
+        # Matches hash to Faiss index
+        self._deduper: t.Dict[str, faiss.IndexFlatL2] = {}
+        # Entry mapping: Each list[entries]'s index is its hash's index
+        self._idx_to_entries: t.List[t.List[IndexT]] = []
+
+        self.add_all(entries=entries)
+
+    def __len__(self) -> int:
+        return len(self._idx_to_entries)
+
+    def query(self, hash: str) -> t.Sequence[PDQIndexMatch[IndexT]]:
+        """
+        Look up entries against the index, up to the max supported distance.
+        """
+        results: t.List[PDQIndexMatch[IndexT]] = []
+        matches_list: t.List[t.List[t.Any]] = self.index.search(
+            queries=[hash], threshold=self.threshold
+        )
+
+        for matches in matches_list:
+            for match_hash, distance in matches:
+                entries = self._idx_to_entries[match_hash]  # Get the Faiss index
+                # Create match objects for each entry
+                results.extend(
+                    PDQIndexMatch(
+                        SignalSimilarityInfoWithIntDistance(distance=int(distance)),
+                        entry,
+                    )
+                    for entry in entries
+                )
+        return results
+
+    def add(self, signal_str: str, entry: IndexT) -> None:
+        self.add_all(((signal_str, entry),))
+
+    def add_all(self, entries: t.Iterable[t.Tuple[str, IndexT]]) -> None:
+        for h, i in entries:
+            existing_faiss_id = self._deduper.get(h)
+            if existing_faiss_id is None:
+                self.index.add([h])
+                self._idx_to_entries.append([i])
+                next_id = len(self._deduper)  # Because faiss index starts from 0 up
+                self._deduper[h] = next_id
+            else:
+                # Since this already exists, we don't add it to Faiss because Faiss cannot handle duplication
+                self._idx_to_entries[existing_faiss_id].append(h)
+
+    def serialize(self, fout: t.BinaryIO) -> None:
+        """
+        Serialize the PDQ index to a binary stream.
+        """
+        fout.write(pickle.dumps(self))
+
+    @classmethod
+    def deserialize(cls, fin: t.BinaryIO) -> "PDQIndex2[IndexT]":
+        """
+        Deserialize a PDQ index from a binary stream.
+        """
+        return pickle.loads(fin.read())
```

Comment:
Q: Since these methods already exist in SignalTypeIndex, do we still need to add them here?